### PR TITLE
Fix #591, adding argmax/min, max/min for lists

### LINKED ISSUE #591: Add generic (arg)max/(arg)min
See [here](https://github.com/MatthewDaggitt/agda-routing/blob/master/RoutingLib/Data/List/Extrema.agda)

### PR DESCRIPTION
**Pull request by MatthewDaggitt**
Fixes #591. If people would like to suggest better names for the proofs, please do. I've added pluses to the names in the same vein as the introduction/elimination proofs in e.g. `Data.List.Relation.Binary.All.Properties`, in case we ever want to introduce elimination proofs. That may be overkill...

### CODE REVIEW

File: src/Data/List/Extrema/Core.agda [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+open Max totalOrder
+open Min totalOrder
+
+private
+
+  variable
+    a : Level
+    A : Set a
+
+  <-transʳ : Trans _≤_ _<_ _<_
```
Comment (JacquesCarette) [inline src/Data/List/Extrema/Core.agda]: It seems to me like the next 3 functions and 4 lemmas belong in `Relation.Binary.Construct.NonStrictToStrict` (or a `.Properties` sub-module).

Comment (MatthewDaggitt) [reply src/Data/List/Extrema/Core.agda]: So the proof of the two functions are indeed from `Relation.Binary.Construct.NonStrictToStrict`. They can't be instantiated further as that module doesn't actually know the order is a partial order. I have toyed in my own library with
adding `Relation.Binary.Construct.NonStrictToStrict.PartialOrder` with these proofs instantiated...

**Comment by gallais** [comment]
> Too bad there are no ways in Agda to take advantage of the inherent symmetry of
the situation to save on code.

wlog (#450) may be useful here?

**Comment by JacquesCarette** [comment]
Well spotted re: wlog!  If not, then it would make sense to really dig into why that is.

**Comment by JacquesCarette** [comment]
@ysharoda should also take a look at this.

**Comment by MatthewDaggitt** [comment]
> wlog (#450) may be useful here?

Sorry, I'm probably being slow but I'm struggling to see where to apply it?

File: src/Data/List/Extrema/Core.agda (hunk 1 above)
Comment (MatthewDaggitt) [reply src/Data/List/Extrema/Core.agda]: Just had a look at these again, and I've changed my mind. The first two functions can't be generalised without creating a new module as stated above. The 4 lemmas have to depend on `f` in order to be of the right shape, so even if you strip `f` out and add the lemmas to `Algebra.Construct.NaturalChoice.Min/Max` then you still basically have to re-create the whole function in order to add `f` back in again.

**Comment by MatthewDaggitt** [comment]
As stated above, I can't see where to apply `wlog` as none of the proofs have multiple internal cases, and I don't think `wlog` can be used to exploit the higher level symmetry between min and max. I'm going to mark this as ready for now, but if someone can see a simplification please do say!

**Comment by gallais** [comment]
> As stated above, I can't see where to apply `wlog`

I haven't looked into it myself. Just reminded people in passing that `wlog`
is one way to deal with the inherent symmetry of a problem. Not sure whether
it applies here.

**Comment by JacquesCarette** [comment]
[And I was quite thankful for the reminder!]  It's not clear to me whether the current `wlog` can handle this particular case of symmetry. I wouldn't hold this up for that - the refactoring of the proofs can happen later, as the external interface that this module provides is fine as is.

**Comment by JacquesCarette** [comment]
[Take another look at WLOG and using it in stdlib](https://trello.com/c/lZXADcwJ/647-take-another-look-at-wlog-and-using-it-in-stdlib)

**Comment by JacquesCarette** [comment]
Didn't realize that Trello would back-link a comment, especially from a private board - sorry for the noise!

**Pull request merged**